optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -60,6 +60,21 @@ public VariableListenerSupport(InnerScoreDirector<Solution_, ?> scoreDirector) {
         supplyMap = new LinkedHashMap<>();
     }
 
+    private VariableListener<Solution_, ?>
+            getVariableListenerAsSupply(ShadowVariableDescriptor<Solution_> shadowVariableDescriptor) {
+        VariableListener<Solution_, ?> variableListener = shadowVariableDescriptor.buildVariableListener(scoreDirector);
+        if (variableListener instanceof Supply) {
+            // Many internal implementations already provide the Supply interface; we need to keep those.
+            return variableListener;
+        }
+        // Otherwise wrap the variable listener to implement Supply.
+        if (variableListener instanceof StatefulVariableListener) {
+            return new SupplyStatefulVariableListener<>((StatefulVariableListener<Solution_, ?>) variableListener);
+        } else {
+            return new SupplyVariableListener<>(variableListener);
+        }
+    }
```

Comment:
If we get rid of the Supply interface, we don't need to do this. Not sure if we want to though. (Discuss during meeting)